addons/udes_advanced_packaging/tests/test_package_reservation.py

```diff
@@ -0,0 +1,172 @@
+# -*- coding: utf-8 -*-
+from odoo.exceptions import ValidationError
+from odoo.addons.udes_stock.tests.common import BaseUDES
+
+
+class TestPackageReservation(BaseUDES):
+    @classmethod
+    def setUpClass(cls):
+        super(TestPackageReservation, cls).setUpClass()
+        # enable full package reservation
+        cls.picking_type_pick.u_reserve_as_packages = True
+
+    def test_reserve_full_package_one_product(self):
+        """Test to reserve a full package for one product
+        when initial demand < package quantity
+        """
+
+        # create a package
+        package = self.create_package()
+
+        # create a quant of 4 apples in a stock sublocation for the package
+        quant = self.create_quant(
+            self.apple.id, self.test_stock_location_01.id, 4, package_id=package.id
+        )
+        # the quant should be unreserved
+        self.assertEqual(quant.reserved_quantity, 0, "The quant is reserved")
+
+        # create a Pick for 1 apple
+        create_info = [{"product": self.apple, "uom_qty": 1}]
+        picking = self.create_picking(
+            self.picking_type_pick, products_info=create_info, confirm=True, assign=False
+        )
+        picking.action_assign()
+
+        # the quant should be fully reserved
+        self.assertEqual(quant.reserved_quantity, 4)
+
+        # apple stock.move.lines of the picking should have the package
+        apple_mls = picking.mapped("move_line_ids")
+        self.assertTrue(all([ml.package_id == package for ml in apple_mls]))
+
+    def test_reserve_full_package_two_products(self):
```

Comment:
Can we add "do_unreserve" at the end of the test and check that the state of the picking is the same as before calling action_assign please? In this case, we are creating a move for strawberries that after unreserving won't go, which means we need a tidy up function to clear extra created moves that have now quantity 0.